server/sonar-server/src/main/java/org/sonar/server/computation/step/FillMeasuresWithVariationsStep.java

```diff
@@ -0,0 +1,234 @@
+ * This step MUST be executed after all steps that create some measures
+ *
+ * Note that measures on developer are not handle yet.
+ */
+public class FillMeasuresWithVariationsStep implements ComputationStep {
+
+  private final DbClient dbClient;
+  private final TreeRootHolder treeRootHolder;
+  private final PeriodsHolder periodsHolder;
+  private final MetricRepository metricRepository;
+  private final MeasureRepository measureRepository;
+
+  private final Function<PastMeasureDto, MeasureKey> pastMeasureToMeasureKey;
+
+  public FillMeasuresWithVariationsStep(DbClient dbClient, TreeRootHolder treeRootHolder, PeriodsHolder periodsHolder, final MetricRepository metricRepository,
+    MeasureRepository measureRepository) {
+    this.dbClient = dbClient;
+    this.treeRootHolder = treeRootHolder;
+    this.periodsHolder = periodsHolder;
+    this.metricRepository = metricRepository;
+    this.measureRepository = measureRepository;
+
+    this.pastMeasureToMeasureKey = new Function<PastMeasureDto, MeasureKey>() {
+      @Nullable
+      @Override
+      public MeasureKey apply(@Nonnull PastMeasureDto input) {
+        Metric metric = metricRepository.getById(input.getMetricId());
+        return new MeasureKey(metric.getKey(), input.getCharacteristicId(), input.getRuleId());
+      }
+    };
+  }
+
+  @Override
+  public void execute() {
+    final DbSession dbSession = dbClient.openSession(false);
+    try {
+      final Iterable<Metric> metrics = FluentIterable.from(metricRepository.getAll()).filter(NumericMetric.INSTANCE);
+      final List<Integer> metricIds = FluentIterable.from(metrics).transform(MetricDtoToMetricId.INSTANCE).toList();
+      new VariationMeasuresVisitor(dbSession, metricIds).visit(treeRootHolder.getRoot());
+    } finally {
+      dbSession.close();
+    }
+  }
+
+  private class VariationMeasuresVisitor extends DepthTraversalTypeAwareVisitor {
+
+    private final DbSession session;
+    private final List<Integer> metricIds;
+
+    public VariationMeasuresVisitor(DbSession session, List<Integer> metricIds) {
+      // measures on files are currently purged, so past measures are not available on files
+      super(Component.Type.DIRECTORY, PRE_ORDER);
+      this.session = session;
+      this.metricIds = metricIds;
+    }
+
+    @Override
+    public void visitAny(Component component) {
+      MeasuresWithVariationRepository measuresWithVariationRepository = new MeasuresWithVariationRepository();
+      for (Period period : periodsHolder.getPeriods()) {
+        List<PastMeasureDto> pastMeasures = dbClient.measureDao().selectByComponentUuidAndProjectSnapshotIdAndMetricIds(session, component.getUuid(), period.getSnapshotId(),
+          metricIds);
+        setVariationMeasures(component, pastMeasures, period.getIndex(), measuresWithVariationRepository);
+      }
+
+      for (MeasureWithVariations measureWithVariations : measuresWithVariationRepository.measures()) {
```

Comment:
side effet of having Measure be an immutable object, it forced you to separate computing the variations from storing them
even better would be to use two private method to emphasis this separation of concern